compliance/collection/nodescanv2/nodescan.go

```diff
@@ -16,5 +19,60 @@ type NodeScan struct {
 
 // Scan scans the current node and returns the results as storage.NodeInventory object
 func (n *NodeScan) Scan(nodeName string) (*storage.NodeInventory, error) {
-	return nil, errors.New("Not implemented")
+	componentsHost, err := nodes.Analyze(nodeName, "/host/", false)
+	log.Info("Finished node inventory /host scan")
+	if err != nil {
+		log.Errorf("Error scanning node /host inventory: %v", err)
+	}
+	log.Infof("Components found under /host: %v", componentsHost)
+	if err != nil {
+		return nil, err
+	}
+
+	var protoComponents *scannerV1.Components
+	if componentsHost != nil {
+		protoComponents = protoComponentsFromScanComponents(componentsHost)
+	}
+	m := &storage.NodeInventory{
+		NodeName:   nodeName,
+		ScanTime:   timestamp.TimestampNow(),
+		Components: protoComponents,
+	}
+	return m, nil
+}
+
+func protoComponentsFromScanComponents(c *nodes.Components) *scannerV1.Components {
+	var components []*scannerV1.RHELComponent
+	// For now, we only care about RHEL components, but this must be extended once we support non-RHCOS
+	if c.CertifiedRHELComponents != nil {
+		components = convertRHELComponents(c.CertifiedRHELComponents)
+	}
+	pc := scannerV1.Components{
+		Namespace:          c.OSNamespace.Name,
+		OsComponents:       nil,
+		RhelComponents:     components,
+		LanguageComponents: nil,
+	}
+	return &pc
+}
+
+func convertRHELComponents(rc *database.RHELv2Components) []*scannerV1.RHELComponent {
+	v1rhelc := make([]*scannerV1.RHELComponent, 0)
+	if rc.Packages == nil {
+		log.Warn("No RHEL packages found in scan result")
+		return nil
+	}
```

Comment:
1. Let's make `nil` check before allocating array.
2. Let's allocate array with the expected size because we have that info.

```suggestion
	if rc.Packages == nil {
		log.Warn("No RHEL packages found in scan result")
		return nil
	}
	v1rhelc := make([]*scannerV1.RHELComponent, len(rc.Packages))
```

Reply:
I have an interesting situation here.
If I initialise `v1rhelc` with the known length, I end up with a panic in the protobuf that serializes this array.
If I roll back the change and init it with 0, the code works without running into said panic.

The stacktrace from a running Compliance:
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x9dc0ec]

goroutine 37 [running]:
github.com/stackrox/scanner/generated/scanner/api/v1.(*RHELComponent).MarshalToSizedBuffer(0x0, {0xc005cac026, 0x426, 0x67b1})
    github.com/stackrox/scanner@v0.0.0-20221213095205-a7f43aa5c05d/generated/scanner/api/v1/component.pb.go:974 +0x2c
github.com/stackrox/scanner/generated/scanner/api/v1.(*Components).MarshalToSizedBuffer(0xc000868c00, {0xc005cac026, 0x67b1, 0x67b1})
    github.com/stackrox/scanner@v0.0.0-20221213095205-a7f43aa5c05d/generated/scanner/api/v1/component.pb.go:861 +0x330
github.com/stackrox/rox/generated/storage.(*NodeInventory).MarshalToSizedBuffer(0xc000b242a0, {0xc005cac026, 0x67b1, 0x67b1})
    github.com/stackrox/rox/generated/storage/node.pb.go:1479 +0x259
github.com/stackrox/rox/generated/internalapi/sensor.(*MsgFromCompliance_NodeInventory).MarshalToSizedBuffer(0x34?, {0xc005cac026, 0x67b1, 0x1806e6e?})
    github.com/stackrox/rox/generated/internalapi/sensor/compliance_iservice.pb.go:1224 +0x3c
github.com/stackrox/rox/generated/internalapi/sensor.(*MsgFromCompliance_NodeInventory).MarshalTo(0x2000000000001?, {0xc005cac026, 0x450812?, 0x67b1})
    github.com/stackrox/rox/generated/internalapi/sensor/compliance_iservice.pb.go:1217 +0x47
github.com/stackrox/rox/generated/internalapi/sensor.(*MsgFromCompliance).MarshalToSizedBuffer(0xc0056b3c00, {0xc005cac000, 0x67d7, 0x67d7})
    github.com/stackrox/rox/generated/internalapi/sensor/compliance_iservice.pb.go:1158 +0x115
github.com/stackrox/rox/generated/internalapi/sensor.(*MsgFromCompliance).Marshal(0xc0059aadf0?)
    github.com/stackrox/rox/generated/internalapi/sensor/compliance_iservice.pb.go:1133 +0x56
google.golang.org/protobuf/internal/impl.legacyMarshal({{}, {0x2425690, 0xc0059aadf0}, {0x0, 0x0, 0x0}, 0x0})
    google.golang.org/protobuf@v1.28.1/internal/impl/legacy_message.go:402 +0xa2
google.golang.org/protobuf/proto.MarshalOptions.marshal({{}, 0x70?, 0x0, 0x0}, {0x0, 0x0, 0x0}, {0x2425690, 0xc0059aadf0})
    google.golang.org/protobuf@v1.28.1/proto/encode.go:166 +0x27b
google.golang.org/protobuf/proto.MarshalOptions.MarshalAppend({{}, 0x0?, 0x99?, 0x0?}, {0x0, 0x0, 0x0}, {0x23f76c0?, 0xc0059aadf0?})
    google.golang.org/protobuf@v1.28.1/proto/encode.go:125 +0x79
github.com/golang/protobuf/proto.marshalAppend({0x0, 0x0, 0x0}, {0x240e4b0?, 0xc0056b3c00?}, 0x0?)
    github.com/golang/protobuf@v1.5.2/proto/wire.go:40 +0xa5
github.com/golang/protobuf/proto.Marshal(...)
    github.com/golang/protobuf@v1.5.2/proto/wire.go:23
google.golang.org/grpc/encoding/proto.codec.Marshal({}, {0x2009900, 0xc0056b3c00})
    google.golang.org/grpc@v1.51.0/encoding/proto/proto.go:45 +0x4e
google.golang.org/grpc.encode({0x7f7bdc1fc830?, 0x37edd70?}, {0x2009900?, 0xc0056b3c00?})
    google.golang.org/grpc@v1.51.0/rpc_util.go:595 +0x44
google.golang.org/grpc.prepareMsg({0x2009900?, 0xc0056b3c00?}, {0x7f7bdc1fc830?, 0x37edd70?}, {0x0, 0x0}, {0x0, 0x0})
    google.golang.org/grpc@v1.51.0/stream.go:1708 +0xd2
google.golang.org/grpc.(*clientStream).SendMsg(0xc0056d1680, {0x2009900?, 0xc0056b3c00})
    google.golang.org/grpc@v1.51.0/stream.go:846 +0xfd
github.com/stackrox/rox/generated/internalapi/sensor.(*complianceServiceCommunicateClient).Send(0x241e310?, 0xc0007fc180?)
    github.com/stackrox/rox/generated/internalapi/sensor/compliance_iservice.pb.go:979 +0x2b
main.manageSendingToSensor({0x241e310, 0xc0007fc180}, {0x23f6c60, 0xc0007f0150}, 0xc0007f81e0)
    github.com/stackrox/rox/compliance/collection/main.go:122 +0x1dd
created by main.main
    github.com/stackrox/rox/compliance/collection/main.go:277 +0x558
```

Reply:
I suppose you've tried the snippet that I suggested which is incorrect (unintentionally). Instead of
```go
make([]*scannerV1.RHELComponent, len(rc.Packages))
```
it should be
```go
make([]*scannerV1.RHELComponent, 0, len(rc.Packages))
```

The difference is that the second version reserves capacity, whereas the first version declares the slice with `len(rc.Packages)` zero-values. An attempt to serialize these values leads to an invalid memory read attempt that crashes the process.
Could you please try one more time with the second version?